File: benchmark/comparisons/script/scripts/matmul.py
# matmul.py -- naive O(n^3) square-matrix multiplication, flat 1-D storage.
# argv[1] is the matrix size n. Init outside timed region.

import sys
import time


def make_matrix(n, offset):
    return [float((i + offset) % 17) for i in range(n * n)]


n = int(sys.argv[1])
A = make_matrix(n, 0)
B = make_matrix(n, 7)
C = [0.0] * (n * n)

t0 = time.perf_counter_ns()

for i in range(n):
    for j in range(n):
        s = 0.0
        for k in range(n):
            s += A[i * n + k] * B[k * n + j]
        C[i * n + j] = s

elapsed = time.perf_counter_ns() - t0
print(elapsed)

File: benchmark/comparisons/script/scripts/matmul.lua
-- matmul.lua -- naive O(n^3) square-matrix multiplication, flat 1-D storage.
-- arg[1] is the matrix size n. Init outside timed region.
--
-- Note: Lua tables are 1-indexed by convention, but to keep the index math
-- identical to .pg / .py we use 0-based explicit indices: A[i*n+k].

local function makeMatrix(n, offset)
    local m = {}
    for i = 0, n * n - 1 do
        m[i] = ((i + offset) % 17) * 1.0
    end
    return m
end

local n = tonumber(arg[1])
local A = makeMatrix(n, 0)
local B = makeMatrix(n, 7)
local C = {}
for i = 0, n * n - 1 do C[i] = 0.0 end

local t0 = os.clock()

for i = 0, n - 1 do
    for j = 0, n - 1 do
        local s = 0.0
        for k = 0, n - 1 do
            s = s + A[i * n + k] * B[k * n + j]
        end
        C[i * n + j] = s
    end
end

local elapsed_ns = (os.clock() - t0) * 1e9
print(math.floor(elapsed_ns))
